File: 0108.convert_sorted_array_to_binary_search_tree/solution.java
 public class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode(int x) { val = x; }
 }


class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return sortedArray(nums, 0, nums.length-1);
    }
    
    public TreeNode sortedArray(int[] nums, int start, int end){
        if (start > end) return null;
        if (start == end) return new TreeNode(nums[start]);
        int mid = (start + end) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = sortedArray(nums, start, mid-1);
        root.right = sortedArray(nums, mid+1, end);
        return root;
    }
}

File: 0108.convert_sorted_array_to_binary_search_tree/solution.py
from typing import List


from utils import TreeNode


class Solution:
    def sorted_array_to_BST(self, nums: List[int]) -> TreeNode:
        if not nums:
            return
        # 通过中序遍历可知，根节点所在数组位置的中间
        mid = len(nums) // 2
        root = TreeNode(mid)
        root.left = self.sorted_array_to_BST(nums[0:mid])
        root.right = self.sorted_array_to_BST(nums[mid + 1 : len(nums)])
        return root
